File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        StringBuilder longest = new StringBuilder();
        int left = 0;
        int maxSize = 0;

        for(int right=0; right<s.length(); right++) {
            longest.append(s.charAt(right));

            while(!isUnique(s.substring(left,right+1))) {
                longest.deleteCharAt(0);
                left++;
            }
            maxSize = longest.length() > maxSize ? longest.length() : maxSize;
        }

        return maxSize;
    }

    public boolean isUnique(String str) {
        Set<Character> set = new HashSet<>();
        List<Character> chars = new ArrayList<>();

        for (char x : str.toCharArray()) {
            set.add(x);
            chars.add(x);
        }

        return set.size() == chars.size();
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.ts
function lengthOfLongestSubstring(s: string): number {
    let maxSize: number = 0;
    let left: number = 0;
    let current: string = "";

    for (let right: number = 0; right<s.length; right++) {
        current += s.charAt(right);

        while(!isUnique(s, left, right+1)) {
            current = current.slice(1);
            left++;
        }
        maxSize = current.length > maxSize ? current.length : maxSize;
    }
    return maxSize;
};

function isUnique(str: string, left: number, right: number): boolean {
    const substr = str.substring(left,right);
    const set: Set<string> = new Set(substr.split(''));
    return set.size == substr.length;
}